pkg/storage/ingest/writer.go

```diff
@@ -189,23 +194,21 @@ func (w *Writer) WriteSync(ctx context.Context, partitionID int32, userID string
 // or an error occurred.
 func (w *Writer) produceSync(ctx context.Context, client *kgo.Client, records []*kgo.Record) kgo.ProduceResults {
 	var (
-		remaining = atomic.NewInt64(int64(len(records)))
-		done      = make(chan struct{})
-		resMx     sync.Mutex
-		res       kgo.ProduceResults
+		remaining        = atomic.NewInt64(int64(len(records)))
+		done             = make(chan struct{})
+		resMx            sync.Mutex
+		res              kgo.ProduceResults
+		maxBufferedBytes = int64(w.kafkaCfg.ProducerMaxBufferedBytes)
 	)
 
 	w.writeRequestsTotal.Add(float64(len(records)))
 
 	for _, record := range records {
-		// We use a new context to avoid that other TryProduce() may be cancelled when this call's context is
-		// canceled. It's important to note that cancelling the context passed to TryProduce() doesn't actually
-		// prevent the data to be sent over the wire (because it's never removed from the buffer) but in some
-		// cases may cause all requests to fail with context cancelled.
-		//
-		// We use TryProduce() instead of Produce() so that it will fast fail if the produce buffer is full.
-		// If we would use Produce(), the Produce() function call will block until the buffer can accept the record.
-		client.TryProduce(context.WithoutCancel(ctx), record, func(r *kgo.Record, err error) {
+		promise := func(r *kgo.Record, err error) {
```

Comment:
Can we move this function definition outside of the loop? I can't see any benefit to defining a closure per loop iteration, and instinctively I'd think it's not performant.

Reply:
Good point, done in https://github.com/grafana/mimir/pull/8738/commits/23f046b4e3102dede7faad664b5b2c1715c50b67. 

P.S. In practice there's only 1 record most of the times, but better outside of the loop given there's no point in keeping in the loop.

Reply:
Yeah, the closure in the loop makes you wonder whether it's loop dependent.